### PR TITLE
[#172407010] Notification Hub Services externalization

### PR DESCRIPTION
**Pull request by AleDore**
Added Notification Hub flow skeleton and improved SDK version for performance purpose

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -47,3 +47,96 @@ __pycache__/
 coverage
 
 yarn-error.log
+generated/backend/AcceptedTosVersion.ts
```
Comment (gunzip) [inline .gitignore:50]: generated/**/* is enough

File: NHCallOrchestrator/handler.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import * as t from "io-ts";
```
Comment (gunzip) [inline NHCallOrchestrator/handler.ts:1]: the function (file name) should describe the operation (ie. HandleNotificationCalls, a verb)

File: NHCallService/__test__/index.test.ts [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+import { FiscalCodeHash, index, NotificationHubMessage } from "..";
+import { PlatformEnum } from "../../generated/backend/Platform";
+
+process.env = {
+  APPINSIGHTS_INSTRUMENTATIONKEY: "foo",
+  QueueStorageConnection: "foobar",
+  SPID_BLOB_CONTAINER_NAME: "spidblob",
+  SPID_BLOB_STORAGE_CONNECTION_STRING: "foobar",
+  SPID_LOGS_PUBLIC_KEY: `-----BEGIN PUBLIC KEY-----
```
Comment (gunzip) [inline NHCallService/__test__/index.test.ts:12]: do we need this ?

File: NHCallService/index.ts [hunk 4]
```diff
@@ -0,0 +1,68 @@
+import { Context } from "@azure/functions";
+import * as df from "durable-functions";
+import * as t from "io-ts";
+import { PatternString } from "italia-ts-commons/lib/strings";
+import { Platform } from "../generated/backend/Platform";
+import { OrchestratorInput as NHCallOrchestratorInput } from "../NHCallOrchestrator/handler";
+import { initTelemetryClient } from "../utils/appinsights";
+
+export type FiscalCodeHash = t.TypeOf<typeof FiscalCodeHash>;
+export const FiscalCodeHash = PatternString("[0-9a-f]{64}");
```
Comment (gunzip) [inline NHCallService/index.ts:10]: use a simple 'NonEmptyString' type since the implementation may change

File: NHCallService/index.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+import { Platform } from "../generated/backend/Platform";
+import { OrchestratorInput as NHCallOrchestratorInput } from "../NHCallOrchestrator/handler";
+import { initTelemetryClient } from "../utils/appinsights";
+
+export type FiscalCodeHash = t.TypeOf<typeof FiscalCodeHash>;
+export const FiscalCodeHash = PatternString("[0-9a-f]{64}");
+
+export const NotificationHubNotifyMessage = t.interface({
+  installationId: FiscalCodeHash,
+  kind: t.literal("NotifyKind"),
```
Comment (gunzip) [inline NHCallService/index.ts:14]:
```suggestion
  kind: t.literal("notify"),

```

File: NHCallService/index.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    message_id: t.string,
+    title: t.string
+  })
+});
+export type NotificationHubNotifyMessage = t.TypeOf<
+  typeof NotificationHubNotifyMessage
+>;
+export const NotificationHubCreateOrUpdateMessage = t.interface({
+  installationId: FiscalCodeHash,
+  kind: t.literal("CreateOrUpdateKind"),
```
Comment (gunzip) [inline NHCallService/index.ts:26]:
```suggestion
  kind: t.literal("CreateOrUpdate"),

```

File: NHCallService/index.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  pushChannel: t.string,
+  tags: t.array(t.string)
+});
+export type NotificationHubCreateOrUpdateMessage = t.TypeOf<
+  typeof NotificationHubCreateOrUpdateMessage
+>;
+
+export const NotificationHubDeleteMessage = t.interface({
+  installationId: FiscalCodeHash,
+  kind: t.literal("DeleteKind")
```
Comment (gunzip) [inline NHCallService/index.ts:37]:
```suggestion
  kind: t.literal("Delete")

```

File: NHCallServiceActivity/handler.ts [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  const failure = failActivity(context, logPrefix);
+  const errorOrMessage = ActivityInput.decode(input);
+  if (isLeft(errorOrMessage)) {
+    failure(
+      "Error decoding activity input",
+      readableReport(errorOrMessage.value)
+    );
+  }
+  const message = errorOrMessage.value;
+  if (NotificationHubCreateOrUpdateMessage.is(message)) {
```
Comment (gunzip) [inline NHCallServiceActivity/handler.ts:75]: since the types are mutually exclusive a `switch (message.kind)` is better

File: package.json [hunk 9]
```diff
@@ -12,6 +12,9 @@
     "test": "jest",
     "test:coverage": "jest --coverage",
     "lint": "tslint -p .",
+    "generate:proxy-models": "npm-run-all generate:proxy:*",
```
Comment (gunzip) [inline package.json:15]:
```suggestion
    "generate:proxy-models": "npm-run-all generate:proxy:*",
```

proxy was the old name of io-backend, use 'backend' instead

File: utils/notificationhub.ts [hunk 10]
```diff
@@ -0,0 +1,175 @@
+import * as azure from "azure-sb";
```
Comment (gunzip) [inline utils/notificationhub.ts:1]: merge this into notifications.ts as well

File: utils/notificationhub.ts [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+            error == null
+              ? resolve(successNH())
+              : reject(
+                  `Error while sending notification to NotificationHub| ${error.message}`
+                )
+        )
+      );
+    },
+    errs =>
+      new Error(`Error while sending notification to NotificationHub| ${errs}`)
```
Comment (gunzip) [inline utils/notificationhub.ts:104]:
```suggestion
      new Error(`Error while sending notification to NotificationHub|${errs}`)
```

File: utils/notificationhub.ts [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+            headers: {
+              ["apns-push-type"]: APNSPushType.ALERT,
+              ["apns-priority"]: 10
+            }
+          },
+          (error, _) =>
+            error == null
+              ? resolve(successNH())
+              : reject(
+                  `Error while sending notification to NotificationHub| ${error.message}`
```
Comment (gunzip) [inline utils/notificationhub.ts:98]:
```suggestion
                  `Error while sending notification to NotificationHub|${error.message}`
```

File: HandleNHNotificationCall/function.json [hunk 13]
```diff
@@ -0,0 +1,18 @@
+{
+  "bindings": [
+    {
+      "queueName": "nhcalls",
```
Comment (gunzip) [inline HandleNHNotificationCall/function.json:4]:
```suggestion
      "queueName": "notificationhubcalls",
```

File: HandleNHNotificationCall/function.json [hunk 14]
```diff
@@ -0,0 +1,18 @@
...
+      "connection": "NHStorageConnection",
```
Comment (gunzip) [inline HandleNHNotificationCall/function.json:5]:
```suggestion
      "connection": "NotificationHubCallsStorageConnection",
```

File: package.json [hunk 15]
```diff
@@ -30,11 +34,14 @@
     "danger-plugin-digitalcitizenship": "^0.3.1",
     "fast-check": "^1.16.0",
     "italia-tslint-rules": "^1.1.3",
+    "italia-utils": "^4.1.0",
     "jest": "^24.8.0",
     "jest-mock-express": "^0.1.1",
     "lolex": "^5.1.1",
     "modclean": "^3.0.0-beta.1",
     "npm-run-all": "^4.1.5",
+    "rimraf": "^2.6.2",
```
Comment (gunzip) [inline package.json:43]: you can use shx rm -rf and remove rimraf from deps

File: HandleNHNotificationCall/index.ts [hunk 16]
```diff
@@ -0,0 +1,63 @@
...
+    message_id: t.string,
+    title: t.string
+  })
+});
+export type NotificationHubNotifyMessage = t.TypeOf<
+  typeof NotificationHubNotifyMessage
+>;
+export const NotificationHubCreateOrUpdateMessage = t.interface({
+  installationId: NonEmptyString,
+  kind: t.literal("CreateOrUpdate"),
```
Comment (francescopersico) [inline HandleNHNotificationCall/index.ts:22]:
```suggestion
  kind: t.literal("CreateOrUpdateInstallation"),

```

File: HandleNHNotificationCall/index.ts [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+  payload: t.interface({
+    message: t.string,
+    message_id: t.string,
+    title: t.string
+  })
+});
+export type NotificationHubNotifyMessage = t.TypeOf<
+  typeof NotificationHubNotifyMessage
+>;
+export const NotificationHubCreateOrUpdateMessage = t.interface({
```
Comment (francescopersico) [inline HandleNHNotificationCall/index.ts:20]: CreateOrUpdate what?
CreateOrUpdateInstallation?

File: HandleNHNotificationCall/index.ts [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+  kind: t.literal("CreateOrUpdate"),
+  platform: Platform,
+  pushChannel: t.string,
+  tags: t.array(t.string)
+});
+export type NotificationHubCreateOrUpdateMessage = t.TypeOf<
+  typeof NotificationHubCreateOrUpdateMessage
+>;
+
+export const NotificationHubDeleteMessage = t.interface({
```
Comment (francescopersico) [inline HandleNHNotificationCall/index.ts:31]: Same here

File: HandleNHNotificationCallOrchestrator/handler.ts [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+      `${logPrefix}|Error decoding input|ERROR=${readableReport(
+        errorOrNHCallOrchestratorInput.value
+      )}`
+    );
+    return false;
+  }
+
+  const nhCallOrchestratorInput = errorOrNHCallOrchestratorInput.value;
+
+  yield context.df.callActivity(
```
Comment (francescopersico) [inline HandleNHNotificationCallOrchestrator/handler.ts:50]: You create retryOptions but than use callActivity and not callActivityWithRetry.

Comment (AleDore) [reply HandleNHNotificationCallOrchestrator/handler.ts:50]: You're right !

File: HandleNHNotificationCallOrchestrator/__tests__/handler.test.ts [hunk 20]
```diff
@@ -0,0 +1,88 @@
+/* tslint:disable:no-any */
+// tslint:disable-next-line: no-object-mutation
+process.env = {
+  ...process.env,
+  AZURE_NH_ENDPOINT:
+    "Endpoint=sb://io-lab-notification-hub-ns.servicebus.windows.net/;SharedAccessKeyName=DefaultFullSharedAccessSignature;SharedAccessKey=C4xIzNZv4VrUnu5jkmPH635MApRUj8wABky8VfdPLqg=",
```
Comment (gunzip) [inline HandleNHNotificationCallOrchestrator/__tests__/handler.test.ts:6]: hope it's not a working connection string :)

File: HandleNHNotificationCall/index.ts (hunk 17 above)
Comment (AleDore) [reply HandleNHNotificationCall/index.ts]: Updated!

File: HandleNHNotificationCall/index.ts (hunk 18 above)
Comment (AleDore) [reply HandleNHNotificationCall/index.ts]: Updated

File: HandleNHNotificationCallOrchestrator/__tests__/handler.test.ts (hunk 20 above)
Comment (AleDore) [reply HandleNHNotificationCallOrchestrator/__tests__/handler.test.ts:6]: Nope it's not working right @BurnedMarshal ?

Comment (BurnedMarshal) [reply HandleNHNotificationCallOrchestrator/__tests__/handler.test.ts:6]: I don't know, that is the key that provides to me @gunzip for testing purpose

Comment (gunzip) [reply HandleNHNotificationCallOrchestrator/__tests__/handler.test.ts:6]: it's a real service, plase remove it :)

File: utils/notificationhub.ts (hunk 12 above)
Comment (raicastino) [reply utils/notificationhub.ts]: @AleDore I think that the problem with the execution of the Unit Test could be this line in the Azure DevOps pipeline (it only installs the node module dependencies without making the build): 

https://github.com/pagopa/io-functions-app/blob/90e5a59bc1c19982b124d01a2b42d19a4ba26842/azure-pipelines.yml#L161

with the last changes the tests to be executed need the tsc generation? Try changing the above line with `make: build`...

**Comment by raicastino** [comment]
@AleDore I think that the problem with the execution of the Unit Test could be this line in the Azure DevOps pipeline (it only installs the node module dependencies without making the build): 

https://github.com/pagopa/io-functions-app/blob/90e5a59bc1c19982b124d01a2b42d19a4ba26842/azure-pipelines.yml#L161

with the last changes the tests to be executed need the tsc generation? Try changing the above line with `make: build`...

File: package.json [hunk 21]
```diff
@@ -3,7 +3,7 @@
   "version": "1.13.0",
   "license": "MIT",
   "scripts": {
-    "build": "tsc",
+    "build": "yarn generate:backend-models && tsc",
```
Comment (gunzip) [inline package.json:6]:
```suggestion
   "prebuild": "yarn generate:backend-models",
   "build": "tsc",
```

File: package.json [hunk 22]
```diff
@@ -12,16 +12,21 @@
...
+    "generate:backend-models": "npm-run-all generate:backend:*",
+    "generate:backend:api-models": "shx rm -rf generated/backend && shx mkdir -p generated/backend && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-backend/master/api_backend.yaml --out-dir generated/backend",
+    "generate:backend:notification-models": "shx rm -rf generated/notifications && shx mkdir -p generated/notifications && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-backend/master/api_notifications.yaml --out-dir generated/notifications",
+    "generate:backend:notification-hub": "shx rm -rf generated/notifications && shx mkdir -p generated/notifications && gen-api-models --api-spec https://raw.githubusercontent.com/pagopa/io-backend/172407010-notificatonhub-queue-storage/api_notificationhub_messages.yaml --out-dir generated/notifications",
     "deploy": "npm run build && func azure functionapp publish agid-functions-app-test",
     "dist:modules": "modclean -r -n default:safe && yarn install --production",
-    "predeploy": "npm-run-all build dist:modules"
+    "predeploy": "npm-run-all generate:backend-models build dist:modules"
```
Comment (gunzip) [inline package.json:21]:
```suggestion
    "predeploy": "npm-run-all build dist:modules"
```

File: package.json [hunk 23]
```diff
@@ -3,7 +3,7 @@
   "version": "1.13.0",
   "license": "MIT",
   "scripts": {
-    "build": "tsc",
+    "build": "yarn generate:backend-models && tsc",
     "build:production": "npm run prestart && npm prune --production",
```
Comment (gunzip) [inline package.json:7]:
```suggestion
```

**Comment by gunzip** [comment]
why draft?

**Comment by AleDore** [comment]
> why draft?

Do we have to add reply logic on installation failure or we have to include it in an other PR?

**Comment by gunzip** [comment]
another pr

**Comment by AleDore** [comment]
> another pr

Ok let's review!

**Pull request merged**